src/travel-stipend-calculator.ts

```diff
@@ -0,0 +1,392 @@
+import { parse } from "csv-parse/sync";
+import fs from "fs";
+import path from "path";
+import { createHashKey, PersistentCache } from "./utils/cache";
+import {
+  BASE_LOCAL_TRANSPORT_PER_DAY,
+  BASE_LODGING_PER_NIGHT,
+  BASE_MEALS_PER_DAY,
+  BUSINESS_ENTERTAINMENT_PER_DAY,
+  COST_PER_KM,
+  DEFAULT_TICKET_PRICE,
+  ORIGIN,
+  POST_CONFERENCE_DAYS,
+  PRE_CONFERENCE_DAYS,
+  WEEKEND_RATE_MULTIPLIER,
+} from "./utils/constants";
+import { loadCoordinatesData } from "./utils/coordinates";
+import { getCostOfLivingFactor, loadCostOfLivingData } from "./utils/cost-of-living";
+import { calculateDateDiff, generateFlightDates } from "./utils/dates";
+import { getDistanceKmFromCities } from "./utils/distance";
+import { calculateFlightCost, lookupFlightPrice } from "./utils/flights";
+import { calculateLocalTransportCost } from "./utils/taxi-fares";
+import { Conference, Coordinates, StipendBreakdown } from "./utils/types";
+
+// Initialize caches
+const distanceCache = new PersistentCache<number>("fixtures/cache/distance-cache.json");
+const coordinatesCache = new PersistentCache<Coordinates>("fixtures/cache/coordinates-cache.json");
+const costOfLivingCache = new PersistentCache<number>("fixtures/cache/col-cache.json");
+const stipendCache = new PersistentCache<StipendBreakdown>("fixtures/cache/stipend-cache.json");
+
+// Load the city coordinates mapping
+console.log("Loading city coordinates data...");
+const cityCoordinates = loadCoordinatesData("fixtures/coordinates.csv");
+
+// Load the cost-of-living mapping
+console.log("Loading cost-of-living data...");
+const costOfLivingMapping = loadCostOfLivingData("fixtures/cost_of_living.csv");
+
+// Helper function to calculate flight cost
+async function calculateFlightCostForConference(
+  destination: string,
+  flightDates: { outbound: string; return: string },
+  distanceKm: number,
+  isOriginCity: boolean
+): Promise<number> {
+  // If destination is the same as origin, no flight cost
+  if (isOriginCity) {
+    console.log(`No flight cost for ${destination} (same as origin)`);
+    return 0;
+  }
+
+  // Try to get flight cost from API first
+  const apiFlightPrice = await lookupFlightPrice(destination, flightDates);
```

Comment:
to avoid spamming errors and unnecessary computations check if the API key is set 
```suggestion
  // Try to get flight cost from API first
  let apiFlightPrice: number | null = null;
  if (process.env.SERPAPI_API_KEY) {
    apiFlightPrice = await lookupFlightPrice(destination, flightDates);
  } else {
    console.warn("SERPAPI_API_KEY is not set. Using the enhanced flight cost calculation model.");
  }
```